### PR TITLE
[minor] Add access token

### PR DESCRIPTION
**Pull request by kyfujisa**
TODO
- [x] access token validate implimatation
- [x] test code
- [x] update document

### CODE REVIEW

File: authorizerd.go [hunk 1]
```diff
@@ -248,6 +249,37 @@ func (a *authorizer) VerifyRoleJWT(ctx context.Context, tok, act, res string) er
...
+// VerifyAccessToken verifies the access token for specific resource and return and verification error.
+func (a *authorizer) VerifyAccessToken(ctx context.Context, tok, act, res string, cert *x509.Certificate) error {
+	// TODO:functionalize
+	if act == "" || res == "" {
+		return errors.Wrap(ErrInvalidParameters, "empty action / resource")
+	}
+	// check if exists in verification success cache
+	_, ok := a.cache.Get(tok + act + res)
+	if ok {
+		glg.Debugf("use cached result. tok: %s, act: %s, res: %s", tok, act, res)
```
Comment (golangcibot) [inline authorizerd.go]: Error return value of `glg.Debugf` is not checked (from `errcheck`)

File: authorizerd.go [hunk 2]
```diff
@@ -248,6 +249,37 @@ func (a *authorizer) VerifyRoleJWT(ctx context.Context, tok, act, res string) er
...
+		return nil
+	}
+
+	ac, err := a.roleProcessor.ParseAndValidateAccessToken(tok, cert)
+	if err != nil {
+		glg.Debugf("error parse and validate access tokenn, err: %v", err)
```
Comment (golangcibot) [inline authorizerd.go]: Error return value of `glg.Debugf` is not checked (from `errcheck`)

File: authorizerd.go [hunk 3]
```diff
@@ -248,6 +249,37 @@ func (a *authorizer) VerifyRoleJWT(ctx context.Context, tok, act, res string) er
...
+		return errors.Wrap(err, "error verify access token")
+	}
+	domain := ac.Audience
+	roles := ac.Scope
+
+	// TODO:functionalize
+	if err := a.policyd.CheckPolicy(ctx, domain, roles, act, res); err != nil {
+		glg.Debugf("error check, err: %v", err)
```
Comment (golangcibot) [inline authorizerd.go]: Error return value of `glg.Debugf` is not checked (from `errcheck`)

**Comment by WindzCUHK** [comment]
plx add this file to mute golangci
https://github.com/yahoojapan/athenz-client-sidecar/blob/master/.golangci.yml

**Comment by WindzCUHK** [comment]
please also add unit test for invalid json type, e.g. 
1. `cnf: ""`
1. `cnf: {"x5t#S256": {}}`

File: role/processor_test.go [hunk 4]
```diff
@@ -559,3 +565,431 @@ func Test_rtp_keyFunc(t *testing.T) {
...
 	}
 }
+
+func Test_rtp_ParseAndValidateAccessToken(t *testing.T) {
+	type fields struct {
+		pkp                                   pubkey.Provider
+		jwkp                                  jwk.Provider
+		enableMTLSCertificateBoundAccessToken bool
+		clientCertificateGoBackSeconds        int64
+		clientCertificateOffsetSeconds        int64
```
Comment (golangcibot) [inline role/processor_test.go]: field `clientCertificateOffsetSeconds` is unused (from `unused`)

File: role/processor_test.go [hunk 5]
```diff
@@ -559,3 +565,431 @@ func Test_rtp_keyFunc(t *testing.T) {
 		})
 	}
 }
+
+func Test_rtp_ParseAndValidateAccessToken(t *testing.T) {
+	type fields struct {
+		pkp                                   pubkey.Provider
+		jwkp                                  jwk.Provider
+		enableMTLSCertificateBoundAccessToken bool
+		clientCertificateGoBackSeconds        int64
```
Comment (golangcibot) [inline role/processor_test.go]: field `clientCertificateGoBackSeconds` is unused (from `unused`)

File: authorizerd.go [hunk 6]
```diff
@@ -293,6 +301,35 @@ func (a *authorizer) verify(ctx context.Context, m mode, tok, act, res string) e
...
 
+// VerifyAccessToken verifies the access token for specific resource and return and verification error.
+func (a *authorizer) VerifyAccessToken(ctx context.Context, tok, act, res string, cert *x509.Certificate) error {
+	if act == "" || res == "" {
+		return errors.Wrap(ErrInvalidParameters, "empty action / resource")
+	}
+	// check if exists in verification success cache
+	_, ok := a.cache.Get(tok + act + res)
+	if ok {
+		glg.Debugf("use cached result. tok: %s, act: %s, res: %s", tok, act, res)
```
Comment (golangcibot) [inline authorizerd.go]: Error return value of `glg.Debugf` is not checked (from `errcheck`)

File: authorizerd.go [hunk 7]
```diff
@@ -293,6 +301,35 @@ func (a *authorizer) verify(ctx context.Context, m mode, tok, act, res string) e
...
+		return nil
+	}
+
+	ac, err := a.roleProcessor.ParseAndValidateAccessToken(tok, cert)
+	if err != nil {
+		glg.Debugf("error parse and validate access tokenn, err: %v", err)
```
Comment (golangcibot) [inline authorizerd.go]: Error return value of `glg.Debugf` is not checked (from `errcheck`)

File: authorizerd.go [hunk 8]
```diff
@@ -293,6 +301,35 @@ func (a *authorizer) verify(ctx context.Context, m mode, tok, act, res string) e
...
+		return errors.Wrap(err, "error verify access token")
+	}
+	domain := ac.Audience
+	roles := ac.Scope
+
+	if err := a.policyd.CheckPolicy(ctx, domain, roles, act, res); err != nil {
+		glg.Debugf("error check, err: %v", err)
```
Comment (golangcibot) [inline authorizerd.go]: Error return value of `glg.Debugf` is not checked (from `errcheck`)

**Comment by WindzCUHK** [comment]
some concerns:
1. role folder, maybe refactor
1. valid(), create new interface, remove duplicated code
1. review new option name

File: role/option_test.go [hunk 9]
```diff
@@ -134,3 +134,160 @@ func TestWithJWKProvider(t *testing.T) {
...
+		checkFunc func(Option) error
+	}{
+		{
+			name: "set success",
+			args: args{
+				b: true,
+			},
+			checkFunc: func(opt Option) error {
+				r := &rtp{}
+				opt(r)
```
Comment (golangcibot) [inline role/option_test.go:155]: Error return value of `opt` is not checked (from `errcheck`)

File: role/option_test.go [hunk 10]
```diff
@@ -134,3 +134,160 @@ func TestWithJWKProvider(t *testing.T) {
...
+		checkFunc func(Option) error
+	}{
+		{
+			name: "set success",
+			args: args{
+				t: "2h",
+			},
+			checkFunc: func(opt Option) error {
+				r := &rtp{}
+				opt(r)
```
Comment (golangcibot) [inline role/option_test.go:189]: Error return value of `opt` is not checked (from `errcheck`)

File: role/option_test.go [hunk 11]
```diff
@@ -134,3 +134,160 @@ func TestWithJWKProvider(t *testing.T) {
...
+			},
+		},
+		{
+			name: "empty value",
+			args: args{
+				t: "",
+			},
+			checkFunc: func(opt Option) error {
+				r := &rtp{}
+				opt(r)
```
Comment (golangcibot) [inline role/option_test.go:203]: Error return value of `opt` is not checked (from `errcheck`)

File: authorizerd_test.go [hunk 12]
```diff
@@ -1203,3 +1203,211 @@ func Test_authorizer_GetPolicyCache(t *testing.T) {
...
+				wantErr: "",
+				checkFunc: func(prov *authorizer) error {
+					_, ok := prov.cache.Get("dummyTokdummyActdummyRes")
+					if !ok {
+						return errors.New("cannot get dummyTokdummyActdummyRes from cache")
+					}
+					return nil
+				},
+			}
+		}(),
```
Comment (WindzCUHK) [inline authorizerd_test.go:1259]:
```suggestion
		func() test {
			now := fastime.Now()
			c := gache.New()
			rm := &TokenMock{
				zatc: &role.ZTSAccessTokenClaim{
					Scope: []string{"role"},
					BaseClaim: role.BaseClaim{
						StandardClaims: jwtgo.StandardClaims{
							Audience: "domain",
						},
					},
				},
				wantErr: nil,
			}
			cm := &PolicydMock{
				CheckPolicyFunc: func(ctx context.Context, domain string, roles []string, action, resource string) error {
					if domain != "domain" || len(roles) != 1 || roles[0] != "role" {
						return errors.New("Audience/Scope mismatch")
					}
					return nil
				},
			}
			return test{
				name: "test verify success",
				args: args{
					ctx: context.Background(),
					tok: "dummyTok",
					act: "dummyAct",
					res: "dummyRes",
				},
				fields: fields{
					policyd:       cm,
					roleProcessor: rm,
					cache:         c,
					cacheExp:      time.Minute,
				},
				wantErr: "",
				checkFunc: func(prov *authorizer) error {
					_, expiry, ok := prov.cache.GetWithExpire("dummyTokdummyActdummyRes")
					if !ok {
						return errors.New("cannot get dummyTokdummyActdummyRes from cache")
					}
					wantExpiry := now.Add(time.Minute).UnixNano()
					if wantExpiry > expiry {
						return fmt.Errorf("cache expiry: got = %v, want: %v", expiry, wantExpiry)
					}
					return nil
				},
			}
		}(),
```

Comment (WindzCUHK) [reply authorizerd_test.go:1259]: `import jwtgo "github.com/dgrijalva/jwt-go"`

File: authorizerd.go [hunk 13]
```diff
@@ -152,13 +168,59 @@ func New(opts ...Option) (Authorizerd, error) {
...
+		verifiers = append(verifiers, rtVerifier)
+	}
+
+	if a.verifyRoleCert {
+		rcVerifier := func(r *http.Request, act, res string) error {
+			return a.VerifyRoleCert(r.Context(), r.TLS.PeerCertificates, act, res)
+		}
+		glg.Info("initVerifiers: added role certificate verifier")
+		verifiers = append(verifiers, rcVerifier)
+	}
```
Comment (WindzCUHK) [inline authorizerd.go:217]: `rcVerifier` should come 1st, like the ATS plugin order?

File: option.go [hunk 14]
```diff
@@ -233,3 +244,56 @@ func WithJwkErrRetryInterval(i string) Option {
...
+func NewATProcessorParam(verifyCertThumbprint bool, certBackdateDur, certOffsetDur string) ATProcessorParam {
+	return ATProcessorParam{
+		verifyCertThumbprint: verifyCertThumbprint,
+		certBackdateDur:      certBackdateDur,
+		certOffsetDur:        certOffsetDur,
+	}
+}
+
+// WithATProcessorParam returns a functional option that new access token processor parameters slice
+func WithATProcessorParam(atpParams []ATProcessorParam) Option {
```
Comment (WindzCUHK) [inline option.go:262]:
```suggestion
func WithATProcessorParam(atpParams ...ATProcessorParam) Option {
```

File: role/processor_test.go [hunk 15]
```diff
@@ -559,3 +598,708 @@ func Test_rtp_keyFunc(t *testing.T) {
...
+					ClientID: "domain.tenant.service",
+					UserID:   "domain.tenant.service",
+					Scope:    []string{"admin", "user"},
+					Confirm:  map[string]string{"x5t#S256": "2jt82f2uM8jE2LMcb4erhhTc-uy1yB1iEyp5MnI5uF4"},
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "verify certificate bound accecss token fail, no confirmation claim",
```
Comment (WindzCUHK) [inline role/processor_test.go:960]: better to add 1 more mis-match test case (non-empty)? although they are the same in code logic

File: role/processor_test.go [hunk 16]
```diff
@@ -559,3 +598,708 @@ func Test_rtp_keyFunc(t *testing.T) {
...
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "verify cert principal success",
+			fields: fields{
+				jwkp: jwk.Provider(func(kid string) interface{} {
+					return LoadRSAPublicKeyFromDisk("./asserts/public.pem")
+				}),
+				enableMTLSCertificateBoundAccessToken: true,
```
Comment (WindzCUHK) [inline role/processor_test.go:1078]:
```suggestion
```
not used?

File: role/processor_test.go [hunk 17]
```diff
@@ -559,3 +598,708 @@ func Test_rtp_keyFunc(t *testing.T) {
...
+				clientCertificateOffsetSeconds:        3600,
+			},
+			args: args{
+				cert: &x509.Certificate{
+					KeyUsage: x509.KeyUsageDigitalSignature,
+					Subject: pkix.Name{
+						CommonName: "domain.tenant.service",
+					},
+					NotBefore: time.Unix(1585122381+100, 0), // token's IssuedAt + 100
```
Comment (WindzCUHK) [inline role/processor_test.go:1087]: 👍🏻

File: authorizerd_mock_test.go [hunk 18]
```diff
@@ -105,15 +106,20 @@ type TokenMock struct {
 	role.Processor
 	wantErr error
 	rt      *role.Token
-	c       *role.Claim
+	rc      *role.RoleJWTClaim
```
[the remark this answers is not in the document]
Comment (kyfujisa) [reply authorizerd_mock_test.go:109]: 5cd3616d93516052a96cab834640b51923bff331

File: authorizerd_mock_test.go [hunk 19]
```diff
@@ -105,15 +106,20 @@ type TokenMock struct {
...
 	return rm.rt, rm.wantErr
 }
 
-func (rm *TokenMock) ParseAndValidateRoleJWT(cred string) (*role.Claim, error) {
-	return rm.c, rm.wantErr
+func (rm *TokenMock) ParseAndValidateRoleJWT(cred string) (*role.RoleJWTClaim, error) {
+	return rm.rc, rm.wantErr
+}
+
+func (rm *TokenMock) ParseAndValidateAccessToken(cred string, cert *x509.Certificate) (*role.ZTSAccessTokenClaim, error) {
```
[the remark this answers is not in the document]
Comment (kyfujisa) [reply authorizerd_mock_test.go:121]: ccad210054a749253b99133afc63ab30f0ea0b03

[the remark this answers is not in the document]
Comment (kyfujisa) [reply authorizerd_mock_test.go]: 9d1f2bc06191eca31695baa0af8019b6ab9d9eb9

File: authorizerd_test.go [hunk 20]
```diff
@@ -1203,3 +1203,211 @@ func Test_authorizer_GetPolicyCache(t *testing.T) {
...
+		func() test {
+			c := gache.New()
+			c.Set("dummyTokdummyActdummyRes", "dummy")
+			rm := &TokenMock{
+				zatc:    &role.ZTSAccessTokenClaim{},
+				wantErr: nil,
+			}
+			cm := &PolicydMock{}
```
[the remark this answers is not in the document]
Comment (kyfujisa) [reply authorizerd_test.go:1475]: d9d1373fc07ff3f10ef2a06d0820b7709fad6fa8

File: authorizerd.go (hunk 13 above)
Comment (WindzCUHK) [reply authorizerd.go:223]: b33bfa609454bf5a414632ac383c0b3923253d9a

File: authorizerd_test.go (hunk 12 above)
Comment (kyfujisa) [reply authorizerd_test.go:1467]: 5b3d15aafd5aa625858f8338c5e9ec992848dd94

File: option.go (hunk 14 above)
Comment (kyfujisa) [reply option.go]: ab9a507d5ac2019cc2bfb30bdcacbbd32da49a05

File: role/processor.go [hunk 21]
```diff
@@ -25,24 +28,36 @@ import (
 	"github.com/yahoojapan/athenz-authorizer/v2/pubkey"
 )
 
+var (
+	confirmationMethod = "x5t#S256"
+)
```
[the remark this answers is not in the document]
Comment (kyfujisa) [reply role/processor.go]: 4c94f2fa526126a5bbdb1967623040d559eae449

File: role/processor.go [hunk 22]
```diff
@@ -106,6 +121,81 @@ func (r *rtp) validate(rt *Token) error {
...
+
+	return claims, nil
+}
+
+func (r *rtp) validateCertificateBoundAccessToken(cert *x509.Certificate, claims *ZTSAccessTokenClaim) error {
+	if cert == nil {
+		return errors.New("error mTLS client certificate is nil")
+	}
+
+	if _, ok := claims.Confirm[confirmationMethod]; !ok {
```
[the remark this answers is not in the document]
Comment (kyfujisa) [reply role/processor.go]: 5289002fdcf59f6eab1c98b1ca53d2e2fa77b147

File: role/processor.go [hunk 23]
```diff
@@ -106,6 +121,81 @@ func (r *rtp) validate(rt *Token) error {
...
+		return errors.New("error token is not certificate bound access token")
+	}
+
+	// cnf check
+	sum := sha256.Sum256(cert.Raw)
+	if base64.RawURLEncoding.EncodeToString(sum[:]) == claims.Confirm[confirmationMethod] {
```
[the remark this answers is not in the document]
Comment (kyfujisa) [reply role/processor.go]: 5289002fdcf59f6eab1c98b1ca53d2e2fa77b147

File: role/processor.go [hunk 24]
```diff
@@ -106,6 +121,81 @@ func (r *rtp) validate(rt *Token) error {
...
+
+	// Issue time check. If the certificate had been updated, it would have been issued later than the token.
+	if cert.NotBefore.Unix() < claims.IssuedAt-r.clientCertificateGoBackSeconds {
+		return errors.Errorf("error certificate: %v issued before token: %v", cert.NotBefore.Unix(), claims.IssuedAt)
+	}
+	// Issue time check. Determine if certificate's issue time is within an allowed range
+	if cert.NotBefore.Unix() > claims.IssuedAt+r.clientCertificateOffsetSeconds-r.clientCertificateGoBackSeconds {
+		return errors.Errorf("Certificate: %v past configured offset %v for token: %v", cert.NotBefore.Unix(), r.clientCertificateOffsetSeconds, claims.IssuedAt)
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (kyfujisa) [reply role/processor.go:196]: 52bc78be9c77b0395844cf573b1a81c07c7fca03

File: role/processor_test.go [hunk 25]
```diff
@@ -559,3 +598,708 @@ func Test_rtp_keyFunc(t *testing.T) {
...
+					enableMTLSCertificateBoundAccessToken: true,
+				},
+				args: args{
+					cred: `eyJraWQiOiIwIiwidHlwIjoiYXQrand0IiwiYWxnIjoiUlMyNTYifQ.eyJzdWIiOiJkb21haW4udGVuYW50LnNlcnZpY2UiLCJpYXQiOjE1ODUxMjIzODEsImV4cCI6OTk5OTk5OTk5OSwiaXNzIjoiaHR0cHM6Ly96dHMuYXRoZW56LmlvIiwiYXVkIjoiZG9tYWluLnByb3ZpZGVyIiwiYXV0aF90aW1lIjoxNTg1MTIyMzgxLCJ2ZXIiOjEsInNjcCI6WyJhZG1pbiIsInVzZXIiXSwidWlkIjoiZG9tYWluLnRlbmFudC5zZXJ2aWNlIiwiY2xpZW50X2lkIjoiZG9tYWluLnRlbmFudC5zZXJ2aWNlIiwiY25mIjp7Ing1dCNTMjU2IjoiMmp0ODJmMnVNOGpFMkxNY2I0ZXJoaFRjLXV5MXlCMWlFeXA1TW5JNXVGNCJ9fQ.OyotreYeMFDTpDaIoPVnEBY1RnVuzRortfRKnkOfZUEv1wSSmgSPxBE9IfgxD57kCQUJtO4GUBUWX_DrIb8BMMVUaDlws6UTncaCUdTt_lJXuIZilh7vIA5oiRTtpADJrZUS3kH2ln6qTXa1QTeevg5qdfORya7ILiHdJUmQXbb9vndYcS4-4E3Xr7rqj7cD67rvySM8YIOsaMn2UX237VUo2rcs40XuHH6WCFfix4xxmgTxS7zr_uowqxpXrgpc0g_eT4On9gnuTDcAzwVy7qbgWMcEO-UrhV_FiPzIRj5RZFZBeHjNeU2QAAT-LAw7S6YJtlPpijfTM9qx6xC0GA`,
+					cert: nil,
+				},
+				wantErr: true,
+			}
+		}(),
+	}
```
[the remark this answers is not in the document]
Comment (kyfujisa) [reply role/processor_test.go:792]: 6eeac96010c38aab67d4f304a477a1fc11c9f210

File: role/processor_test.go [hunk 26]
```diff
@@ -559,3 +598,708 @@ func Test_rtp_keyFunc(t *testing.T) {
...
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "verify certificate bound accecss token success",
+			fields: fields{
+				jwkp: jwk.Provider(func(kid string) interface{} {
+					return LoadRSAPublicKeyFromDisk("./asserts/public.pem")
+				}),
+				enableMTLSCertificateBoundAccessToken: true,
```
[the remark this answers is not in the document]
Comment (kyfujisa) [reply role/processor_test.go:863]: 43f2cf4acb894e171f7e7efd61d3a7a4368a7cc3

File: role/processor_test.go (hunk 15 above)
Comment (kyfujisa) [reply role/processor_test.go:958]: 930b45d7acc36c92377872b0f6597ff6d7c15c98

File: role/processor_test.go (hunk 16 above)
Comment (kyfujisa) [reply role/processor_test.go:1076]: 43f2cf4acb894e171f7e7efd61d3a7a4368a7cc3

**Comment by kyfujisa** [comment]
https://github.com/yahoojapan/athenz-authorizer/issues/60

File: authorizerd.go [hunk 27]
```diff
@@ -152,13 +168,65 @@ func New(opts ...Option) (Authorizerd, error) {
...
+		glg.Info("initVerifiers: added role token verifier")
+		verifiers = append(verifiers, rtVerifier)
+	}
+
+	if len(verifiers) < 1 {
+		return errors.New("error no verifiers")
+	}
+
+	// resize
+	a.verifiers = append([]verifier(nil), verifiers...)
```
Comment (kpango) [inline authorizerd.go:226]:
```suggestion
	copy(a.verifiers[0:],  verifiers)
```

File: authorizerd.go [hunk 28]
```diff
@@ -152,13 +168,65 @@ func New(opts ...Option) (Authorizerd, error) {
...
+		glg.Info("initVerifiers: added role token verifier")
+		verifiers = append(verifiers, rtVerifier)
+	}
+
+	if len(verifiers) < 1 {
+		return errors.New("error no verifiers")
+	}
+
+	// resize
+	a.verifiers = verifiers
```
Comment (kpango) [inline authorizerd.go:226]: This code is not thread-safe

File: authorizerd.go [hunk 29]
```diff
@@ -152,13 +168,66 @@ func New(opts ...Option) (Authorizerd, error) {
...
+		glg.Info("initVerifiers: added role token verifier")
+		verifiers = append(verifiers, rtVerifier)
+	}
+
+	if len(verifiers) < 1 {
+		return errors.New("error no verifiers")
+	}
+
+	// resize
+	a.verifiers = make([]verifier, len(verifiers), len(verifiers))
```
Comment (golangcibot) [inline authorizerd.go:226]: S1019: should use make([]verifier, len(verifiers)) instead (from `gosimple`)

File: authorizerd.go (hunk 28 above)
Comment (kyfujisa) [reply authorizerd.go]: thanks, 0c73eaa

File: authorizerd.go (hunk 27 above)
Comment (kyfujisa) [reply authorizerd.go]: `a.verifiers` does not have an element.
So I can't use `copy()` as it is (nil is assigned to `a.verifiers`).
So I used make.  0c73eaa

File: authorizerd.go (hunk 29 above)
Comment (kyfujisa) [reply authorizerd.go]: 96a711e

File: authorizerd.go (hunk 27 above)
Comment (kpango) [reply authorizerd.go]: I think https://github.com/yahoojapan/athenz-authorizer/commit/0c73eaaa47d6d54a105fc7160cd7ead65d23af17 looks good but I read your first code, that was bad
```
a.verifiers = append([]verifier(nil), verifiers...)
```
why don't you allocate a.verifiers at L191 ? in this case we don't need to define verifiers

Comment (kyfujisa) [reply authorizerd.go]: understand, thank you.

File: role/processor.go [hunk 30]
```diff
@@ -25,24 +28,36 @@ import (
 	"github.com/yahoojapan/athenz-authorizer/v2/pubkey"
 )
 
+var (
```
Comment (WindzCUHK) [inline role/processor.go:31]:
```suggestion
const (
```

File: role/processor_test.go [hunk 31]
```diff
@@ -559,3 +598,729 @@ func Test_rtp_keyFunc(t *testing.T) {
...
+							Issuer:    "https://zts.athenz.io",
+							Audience:  "domain.provider",
+						},
+					},
+					AuthTime: 1585122381,
+					Version:  1,
+					ClientID: "domain.tenant.service",
+					UserID:   "domain.tenant.service",
+					Scope:    []string{"admin", "user"},
+					//  cnf when cert thumbprint is "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
```
Comment (WindzCUHK) [inline role/processor_test.go:1015]: this comment is not needed?

File: role/processor_test.go [hunk 32]
```diff
@@ -559,3 +598,729 @@ func Test_rtp_keyFunc(t *testing.T) {
...
+			return test{
+				name: "verify certificate bound access token fail, empty cnf (\"cnf\": {})",
+				fields: fields{
+					jwkp: jwk.Provider(func(kid string) interface{} {
+						return LoadRSAPublicKeyFromDisk("./asserts/public.pem")
+					}),
+					enableMTLSCertificateBoundAccessToken: true,
+				},
+				args: args{
+					cred: `eyJraWQiOiIwIiwidHlwIjoiYXQrand0IiwiYWxnIjoiUlMyNTYifQ.eyJzdWIiOiJkb21haW4udGVuYW50LnNlcnZpY2UiLCJpYXQiOjE1ODUxMjIzODEsImV4cCI6OTk5OTk5OTk5OSwiaXNzIjoiaHR0cHM6Ly96dHMuYXRoZW56LmlvIiwiYXVkIjoiZG9tYWluLnByb3ZpZGVyIiwiYXV0aF90aW1lIjoxNTg1MTIyMzgxLCJ2ZXIiOjEsInNjcCI6WyJhZG1pbiIsInVzZXIiXSwidWlkIjoiZG9tYWluLnRlbmFudC5zZXJ2aWNlIiwiY2xpZW50X2lkIjoiZG9tYWluLnRlbmFudC5zZXJ2aWNlIiwiY25mIjp7Ing1dCNTMjU2Ijp7fX19.jil95otB9yKJCUTrpbl5h592HKQYDquhz5vSfeAv3fcmFkvSAIcwizKPlWfW2g3eF_VZVu3WRgJyHCwIfWKrOTqdGGJYJs1QmkDAYhjkpgsLCBVXAM63dBIGKRkpt2eO4biqPfjynE--RcJLK1qMVpoWUaQW0QnM5XjjVpgcJtR1l3sjnUXGO4J2L0Q_CAyHSoQvd00jlWA72c32V2L10mQ5BXtEaaEgC8nVc6sdLtAhfUOvyx7xhRFuWl9VL-OO8Z2_5jx1goF2E6Icejmjk5gtKTJa9Wo-ck-P4stphWbNQEpfia6JJXE3y-LKBvdZKmCJhVpKUY_ZWSOLaNQpgA`,
```
Comment (WindzCUHK) [inline role/processor_test.go:838]: same as token in L820?

File: README.md [hunk 33]
```diff
@@ -55,18 +55,21 @@ Athenz policy daemon (policyd) is responsible for periodically update the policy
...
+| PubkeyEtagFlushDur    | ETag cache purge duration                                                                                                | 84 Hours                                                                                        | No       |                        |
+| PolicyRefreshDuration | The refresh duration to update Athenz policy data                                                                        | 30 Minutes                                                                                      | No       |                        |
+| PolicyExpireMargin    | The expire margin to update the policy data. It forces update the policy data before the policy expiration margin.       | 3 Hours                                                                                         | No       |                        |
+| ATProcessorParam      | List of parameters for verifying access token. See [here](./option.go) for details of the options that can be specified. | verify cert thumbprint: enable, cert backdate duration: 1 hours, cert offset duration: 1 hours. | No       |                        |
```
Comment (WindzCUHK) [inline README.md:70]:
```suggestion
| ATProcessorParams      | List of parameters for verifying access token. See [here](./option.go) for details of the options that can be specified. | verify cert thumbprint: enable, cert backdate duration: 1 hours, cert offset duration: 1 hours. | No       |                        |
```

File: role/processor.go (hunk 30 above)
Comment (kyfujisa) [reply role/processor.go]: 5337be8

File: role/processor_test.go (hunk 31 above)
Comment (kyfujisa) [reply role/processor_test.go]: I thought it would be confusing if there was no comment.
It seems unnecessary, so I deleted it.

c9b254a

File: role/processor_test.go (hunk 32 above)
Comment (kyfujisa) [reply role/processor_test.go]: thank you.
fixed.
cf29d08

File: README.md (hunk 33 above)
Comment (kyfujisa) [reply README.md]: 2b60f79

**Comment by WindzCUHK** [comment]
LGTM~
